src/gui/editors/TimeLineWidget.cpp

```diff
@@ -225,83 +212,138 @@ void TimeLineWidget::paintEvent( QPaintEvent * )
+	{
+		auto color = loopPointsActive ? m_activeLoopHandleColor : m_inactiveLoopHandleColor;
+		p.fillRect(leftHandle, color);
+		p.fillRect(rightHandle, color);
+	}
 
 	// Only draw the position marker if the position line is in view
-	if (m_posMarkerX >= m_xOffset && m_posMarkerX < width() - m_posMarkerPixmap.width() / 2)
+	if (markerX(m_pos) >= m_xOffset && markerX(m_pos) < width() - m_posMarkerPixmap.width() / 2)
 	{
 		// Let the position marker extrude to the left
 		p.setClipping(false);
 		p.setOpacity(0.6);
-		p.drawPixmap(m_posMarkerX, height() - m_posMarkerPixmap.height(), m_posMarkerPixmap);
+		p.drawPixmap(markerX(m_pos) - (m_posMarkerPixmap.width() / 2),
+			height() - m_posMarkerPixmap.height(), m_posMarkerPixmap);
 	}
 }
 
+auto TimeLineWidget::getClickedTime(const int xPosition) const -> TimePos
+{
+	// How far into the timeline we clicked, measuring pixels from the leftmost part of the editor
+	const auto pixelDelta = std::max(xPosition - m_xOffset, 0);
+	return m_begin + static_cast<int>(pixelDelta * TimePos::ticksPerBar() / m_ppb);
+}
 
+auto TimeLineWidget::getLoopAction(QMouseEvent* event) const -> TimeLineWidget::Action
+{
+	const auto mode = ConfigManager::inst()->value("app", "loopmarkermode");
+	const auto xPos = event->x();
+	const auto button = event->button();
 
+	if (mode == "handles")
+	{
+		// Loop start and end pos, or closest edge of screen if loop extends off it
+		const auto leftMost = std::max(markerX(m_timeline->loopBegin()), m_xOffset);
+		const auto rightMost = std::min(markerX(m_timeline->loopEnd()), width());
+		// Distance from click to handle, positive aimed towards center of loop
+		const auto deltaLeft = xPos - leftMost;
+		const auto deltaRight = rightMost - xPos;
+
+		if (deltaLeft < 0 || deltaRight < 0) { return Action::NoAction; } // Clicked outside loop
+		else if (deltaLeft <= m_loopHandleWidth && deltaLeft < deltaRight) { return Action::MoveLoopBegin; }
+		else if (deltaRight <= m_loopHandleWidth) { return Action::MoveLoopEnd; }
+		else { return Action::MoveLoop; }
+	}
+	else if (mode == "closest")
+	{
+		const TimePos loopMid = (m_timeline->loopBegin() + m_timeline->loopEnd()) / 2;
+		return getClickedTime(xPos) < loopMid ? Action::MoveLoopBegin : Action::MoveLoopEnd;
+	}
+	else // Default to dual-button mode
+	{
+		if (button == Qt::LeftButton) { return Action::MoveLoopBegin; }
+		else if (button == Qt::RightButton) { return Action::MoveLoopEnd; }
+		return Action::NoAction;
+	}
+}
 
-void TimeLineWidget::mousePressEvent( QMouseEvent* event )
+auto TimeLineWidget::actionCursor(Action action) const -> QCursor
 {
-	if( event->x() < m_xOffset )
-	{
-		return;
+	switch (action) {
+		case Action::MoveLoop: return Qt::SizeHorCursor;
+		case Action::MoveLoopBegin: return m_cursorSelectLeft;
+		case Action::MoveLoopEnd: return m_cursorSelectRight;
+		// Fall back to normal cursor if no action or action cursor not specified
+		default: return Qt::ArrowCursor;
```

Comment:
What do we think better practice would be: using `default:`, or no default with `return x` outside the switch?

Reply:
I'm not aware of a strong argument either way. I prefer using `default` here, as then it's consistent with the other cases, and has a reading that more closely matches the intent (it literally marks the default value with "default").

A couple of arguments I can think of in favour of putting it outside the switch are:
* Compilers can emit warnings if there is no default and a missing case, which helps catch errors introduced when extending the enum. However, this then requires all possible enumerators to have their own case, which is unnecessarily verbose. Also, I don't want to create a precedent for writing code to be read by tools over humans.
* It ensures that a value is still returned if a case breaks instead of returning. However, compilers will typically warn about this anyway, and in simple cases like this, it's difficult to miss such a mistake.

Reply:
Alright, then we can leave this to code readability preference.